### DISCUSSION TITLE
Datasets Page Mockup

### DISCUSSION THREAD

**Discussion by dszidi** [truenas/webui#6768] [Released Features]
## DATASETS PAGE

Users can think of this new Datasets page as a more focused version of the current Storage page. To keep with our strategy, we wanted this page to focus on data management. We kept the datasets tree table, removed all pool management functionality, and replaced it with more information better suited for data management.

<img width="2168" alt="image" src="https://user-images.githubusercontent.com/29865743/172704597-60bee502-dd90-4a3c-8a48-b05da18ff1f9.png">

### LAYOUT

The main layout consists of two areas

1. Tree Table
2. Details Cards

The tree table on the left is filled with only high level information and the cards on the right show details about the selected row. The tree table rows can be expanded, collapsed or selected. To select a row, simply click on it. The row selection updates the contents of the details cards. By default, the first row is selected on page load.

### TREE TABLE

The tree table contains information about the following.

- Dataset Name and Type (type is represented with an icon)
- Usage to show how much space is in use
- Encryption Status (locked/unlocked/unencrypted)
- Roles
- Activity

<img width="807" alt="image" src="https://user-images.githubusercontent.com/29865743/172704057-a654f906-ade6-46e1-8e38-03af75b7d6bf.png">

Most of these are self explanatory but the new elements here are the roles column and the activity indicators. The purpose of the Roles column is to show how the dataset is being used. We want to indicate to the user if the dataset is used by Apps, VMs or sharing protocols. We're using two types of icons for the shares  To show that a dataset has a share directly attached, we use the network folder with the protocol name on it. The generic share icon signifies that the children of that dataset have some sort of shares attached.

<img width="503" alt="image" src="https://user-images.githubusercontent.com/29865743/172705029-7a798193-d7b4-4fcc-ac29-2c027c4feed4.png">

To show activity on the dataset, the UI shows a spinner with the number of tasks in progress. Hovering over this spinner provides a popup with a breakdown of the running tasks. The type of tasks shown here are:

- ZFS Replication
- Rsync
- Cloud Sync
- Snapshot

<img width="447" alt="image" src="https://user-images.githubusercontent.com/29865743/172704272-2d3b18e7-fb30-4ef5-ab17-29126ec7cd32.png">

### DETAILS CARDS

The details cards are where we show information about the following topics

- Dataset
- ZFS Encryption
- Data Protection (overview of snapshots, replication tasks etc)
- Roles (more detailed explanation about the roles column items)
- Permissions (Unix or ACL)
- Capacity Management (Quotas & Reservations etc)

These cards provide sufficient separation and organize the datapoints with the related actions. We also made a point to help users get to any related pages. If the card presents a datapoint, we try to provide a path to the page that allows the user to make adjustments.

<img width="1011" alt="image" src="https://user-images.githubusercontent.com/29865743/172704497-2b6def16-c486-4603-ba92-21f95cc35633.png">

We feel that overall, the removal of pool management actions on this page is more than compensated for with all the new data we're exposing. With an exclusive focus on data management, users can make more informed decisions about their datasets and their data in general.

With all this said, we're not perfect and we've tried our best to cover all the bases. If you feel there is anything that can be improved with this Datasets Page design, please leave your constructive feedback by responding to this thread. Thank you!

**Comment by stavros-k** [comment]
Awesome! Can't wait to have it live!

**Comment by dak180** [comment]
What does this look like at a width of 450 points?

**Reply by dszidi** [reply]
We are prioritizing large screens for now, but will try to make it usable on smaller devices

**Comment by dszidi** [comment]
A quick update on the Capacity Management card in the details area. The plan here is to have separate sections that show usage, available space and other quotas. There will be slight differences between what is shown for datasets and zvols.

**Dataset Space Management**
- In the usage section (top) we will show a graph representing data written to disk by the dataset and it's descendants. Any configured reservations will be shown separately underneath the graph. 
- In the available section (middle) we show the available number as reported by ZFS along with applicable quotas.


![image](https://user-images.githubusercontent.com/29865743/179563641-be29c95c-0208-4073-ae7a-73d30f5fc2b9.png)



**Zvol Space Management**
- In the usage section (top) we will show a graph representing data written to disk by the volume and snapshots. Underneath the graph we will show volume size and provisioning type (thick/thin).  
- In the available section (middle) we show the available number as reported by ZFS along with inherited quotas. We do not however explicitly show quotas/reservations since that functionality is baked into the provisioning volume size settings

![image](https://user-images.githubusercontent.com/29865743/179563577-00b72eb2-57e7-42f6-81e0-c45fe57fe57e.png)

**Reply by AlexHunterScot** [reply]
I actually quite liked the bar showing the data size, snapshot size and quota relative to the available capacity. Was the bar's total just the pool's total capacity or did it account for parent dataset quotas? How will quotas be represented with the pie chart?

If you have the dataset properties enabled, it would be nice to see compression and dedupe ratios to see how much space they're saving.

**Comment by AlexHunterScot** [comment]
Agree with the decision to separate pool and dataset management so that each page can be more specialised with more data.

This is not feedback specifically about this specific new UI, it's more about TrueNAS' general design approach. There's _so much empty space_, just everywhere in TrueNAS! It's difficult to find things at a glance, skimmability is terrible, and everything requires lots of clicks drilling down to what I want. Nothing jumps out because everything is either a huge table, a list of labels and numbers, or forms with huge spacing and lots of scrolling. It's bad enough that zooming out your browser just to fit everyting on one 1080p window is often necessary, and forget splitting your monitor vertically or using a tablet/phone. Although this new UI looks nice, my problems with it are all examples of systemic UI design problems that are spread throughout TrueNAS. You need to think about those first before polishing specific screens.

For example, I suspect the reason the manage links each repeat the name of the thing they're managing is that without being more descriptive, the links all look the same and it's hard to tell at a glance which one does what. Your solution to that problem means you have to leave more empty space to account for the longer and variable-length links, creating the original problem the solution is meant to be fixing. It's a self-fulfilling prophecy.

Instead, you could bring them closer together, or alternate the background colour of the rows to visually separate them, or replace them with a gear or pencil icon button, or get rid of the manage links entirely and make the field names or values themselves links; or fill the empty space with useful things that help me see and do what I want quicker and with fewer clicks.

Lots of those fields say "0". If I know a dataset has no replication tasks and I've specifically navigated through the UI and looked for that field, I probably already know it has no replication tasks, _specifically because_ I know I want to add one.

This UI doesn't help me do that, because the "Manage Replication Tasks" link will take me to the main Replication Tasks page with the tasks for all datasets, which I can access quicker through the left navigation menu. By using this new UI to add a replication task to my dataset, the new UI has wasted my time with a longer and more complex route to the same place: opening the Replication Tasks screen and clicking Add.

Each replication task has a user-supplied name, so instead of this:

**Replication tasks:** _3_ ________________________ [Manage Replication Tasks]()

why not this:

**Replication tasks:** _[name1](), [name2](), [name3]()_. ____________ [Add]() | [Manage]()

In my quick and dirty GitHub markdown example, they both take up the same amount of horizontal space, yet the 2nd one is so much more useful and easier to visually interpret. If there's not enough horizontal space, the list could wrap within the card's table, or truncate to:

**Replication tasks:** _[name1](), and [2 more]()_

**Reply by undsoft** [reply]
Thanks for the feedback, we'll discuss it internally.

**Comment by robloh** [comment]
I like the overview of the tree, that's very useful. Would this then replace the Shares page as it's already visible on the Datasets page now?

**Reply by undsoft** [reply]
Not at the moment, there are other things you can do in Shares.

**Comment by GoingTime** [comment]
I've found that many people use datasets as folders, which is clearly inappropriate.  

On the dataset page, we can consider adding folder creation and management functions, which will bring some benefits:

1. We can quickly create and manage folders through the graphical interface, no need to enter the CLI operation.
2. Can help first-time contacts  people understand the concept of "dataset" faster.

**Reply by undsoft** [reply]
Yeah, we've considered adding some support for folder management. Maybe this is something we do in the future.
